### DISCUSSION TITLE
Does HomeSpan send a hap notification for every `SpanCharacteristic.setVal()`?

### DISCUSSION THREAD

**Discussion by aaaidan** [HomeSpan/HomeSpan#34] [Q&A] [answered]
Here's a naive loop that keeps a SpanCharacteristic in sync with the state of a pin.

```
SpanCharacteristic* sensorState; // initialized in setup()

void loop() {
  sensorState->setVal( digitalRead(PIN_CONTACT_SENSOR) );
  homeSpan.poll();
}
```

This calls setVal every loop, regardless of whether the pin value has changed. I assumed HomeSpan would check whether the value is actually updated, but is this a bad practice? 

I notice `SpanCharacteristic::setVal()` functions look like they're unconditionally pushing a notification to be serviced by `HAPClient::eventNotify`. Does this mean my little ESP32 is saturating its connection with homekit updates? 🤣

**Comment by HomeSpan** [comment, marked as answer]
HomeSpan does *not* check that the state has changed and will generate event notifications whenever setVal() is called.  And yes, it's not a good idea to continuously send notifications since it does create a LOT of network traffic.  From HomeSpan example 13, which introduces setVal():

>   // NOTE: It it NOT recommended to continuously change Characteristic values using setVal() as this will generate a lot of network traffic since HomeSpan
  // sends Event Notifications bck to all registered HomeKit Controllers.  It's fine to perform internal calculations, generate signals on different pins,
  // and perform any other internal actions you may need as frequently as you require.  But limit the use of setVal() to a reasonable frequency, such as maybe
  // one per minute for a temperature sensor.  Do not use setVal() unless the value of the Characteristic changes, but do use it to immediately inform HomeKit of
  // something time-sensitive, such as a door opening, or a smoke alarm triggering.

**Reply by aaaidan** [reply]
Thank you for this info! It's weird to think my back gate sensor has been continuously streaming junk to the network for the last week. 😬 

I'll wrap my thing in a `Bounce` and only `setVal` when there's a `Bounce.change() == 1`. 😀 

It's probably always good practice to debounce pins, and debouncing usually gives you change tracking for free. But do you think there's room for HomeSpan to compare the `setVal` param with the existing internal value, and bailing if it's equal?

**Comment by HomeSpan** [comment]
It's a good suggestion, but I'd be hesitant to enforce that in setVal() since there may be valid reasons to send the same value under certain circumstances.  For example, if you have an external scene or automation that relies on reading a sensor, but the automation itself drops out and misses a changed value, sending an occasional repeat would allow the automation to re-sync.

So I guess the rule is: don't repeatedly send the same values, but it's okay to send them periodically just to make sure every device listening is in sync.  How often?  Often enough to keep things in sync, but not too often to create lots of network traffic.
